File: CSharp.Data.HList/HList.cs
using System;
using Microsoft.FSharp.Core;

namespace CSharp.Data
{
    public abstract class HList<A> where A : HList<A>
    {
        internal HList() {}

        public abstract HCons<E, A> Extend<E>(E e);
        public abstract Apply<Unit, Tuple<E, A>, HCons<E, A>> Extender<E>();
        public abstract int Length { get; }
    }

    public static class HList
    {
        public static HCons<E, L> Cons<E, L>(E e, L l) where L : HList<L>
        {
            return new HCons<E, L>(e, l);
        }

        private static readonly HNil nil = new HNil();
        public static HNil Nil() { return nil; }

        public static HCons<E, HNil> Single<E>(E e)
        {
            return new HCons<E, HNil>(e, Nil());
        }
    }

    public class HNil : HList<HNil>
    {
        internal HNil() {}

        public override HCons<E, HNil> Extend<E>(E e)
        {
            return HList.Cons(e, this);
        }

        public override Apply<Unit, Tuple<E, HNil>, HCons<E, HNil>> Extender<E>()
        {
            return Apply.Cons<E, HNil>();
        }

        public override int Length { get { return 0; } }
    }

    public class HCons<E, L> : HList<HCons<E, L>> where L : HList<L>
    {
        internal HCons(E e, L l)
        {
            Head = e;
            Tail = l;
            Length = l.Length + 1;
        }

        public override HCons<X, HCons<E, L>> Extend<X>(X e)
        {
            return HList.Cons(e, this);
        }

        public override Apply<Unit, Tuple<X, HCons<E, L>>, HCons<X, HCons<E, L>>> Extender<X>()
        {
            return Apply.Cons<X, HCons<E, L>>();
        }

        public E Head { get; }

        public L Tail { get; }

        public override int Length { get; }
    }
}

File: src/FSharp.Data.HList/HList.fs
namespace FSharp.Data

open CSharp.Data

module HList =

  let nil = HList.Nil()

  let inline cons e (l: #HList<_>) = l.Extend(e)

  let inline singleton e = HList.Single(e)

  let inline head (c: HCons<_, _>) = c.Head
  let inline tail (c: HCons<_, _>) = c.Tail

  let inline append (h: #HAppend<_, _, _>) a b = h.Append(a, b)

  let inline length (h: HList<_>) = h.Length

  let inline foldBack (foldr: HFoldr<_, _, _, _>) f v l = foldr.FoldRight(f, v, l)

module HAppend =

  let define<'T when 'T :> HList<'T>> = HAppend.Append<'T>()

  let inline append (h: HAppend<_, _, _>) = HAppend.Append(h)

module Apply =

  let id<'A> = Apply.Id<'A>()

  let func = { new Apply<_ -> _, _, _> with
    member __.Apply(f, a) = f a }

  let comp = { new Apply<unit, (_ -> _) * (_ -> _), _ -> _> with
    member __.Apply((), (a1, a2)) = a1 >> a2 }

  let cons<'E, 'L when 'L :> HList<'L>> = { new Apply<unit, 'E * 'L, HCons<'E, 'L>> with
    member __.Apply((), (a1, a2)) = HList.cons a1 a2 }

  let append = { new Apply<HAppend<_, _, _>, _ * _, _> with
    member __.Apply(f, (a1, a2)) = f.Append(a1, a2) }

module HFoldBack =

  let define<'G, 'V> = HFoldr.Hfoldr<'G, 'V>()

  let inline foldBack p h = HFoldr.Hfoldr(p, h)

[<AutoOpen>]
module HListSyntax =

  let (+|) e l = HList.cons e l
